sound/soc/intel/boards/ehl_rt5660.c

```diff
@@ -0,0 +1,343 @@
+		.nonatomic = true,
+		SND_SOC_DAILINK_REG(ssp0_pin, rt5660_codec, platform),
+	},
+	{
+		.name = "dmic48k",
+		.id = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
+	{
+		.name = "dmic16k",
+		.id = 2,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic16k, dmic_codec, platform),
+	},
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+#define HDMI_LINK_START 3
+#define HDMI_LINE_END 6
+	{
+		.name = "iDisp1",
+		.id = 5,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
+	},
+	{
+		.name = "iDisp2",
+		.id = 6,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
+	},
+	{
+		.name = "iDisp3",
+		.id = 7,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
+	{
+		.name = "iDisp4",
+		.id = 8,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp4_pin, idisp4_codec, platform),
+	},
+#endif
+};
+
+/* SoC card */
+static struct snd_soc_card snd_soc_card_ehl_rt5660 = {
+	.name = "ehl-rt5660",
+	.owner = THIS_MODULE,
+	.dai_link = ehl_rt5660_dailink,
+	.num_links = ARRAY_SIZE(ehl_rt5660_dailink),
+	.dapm_widgets = rt5660_widgets,
+	.num_dapm_widgets = ARRAY_SIZE(rt5660_widgets),
+	.dapm_routes = rt5660_map,
+	.num_dapm_routes = ARRAY_SIZE(rt5660_map),
+	.controls = rt5660_controls,
+	.num_controls = ARRAY_SIZE(rt5660_controls),
+	.fully_routed = true,
+	.late_probe = card_late_probe,
+};
+
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+/* If hdmi codec is not supported, switch to use dummy codec */
+static void sof_refine_hdmi_links(struct snd_soc_card *card,
+				  struct snd_soc_acpi_mach *mach)
+{
+	struct snd_soc_dai_link *link;
+	int i;
+
+	if (mach->mach_params.common_hdmi_codec_drv &&
+	    (mach->mach_params.codec_mask & IDISP_CODEC_MASK))
+		return;
+
+	/* hdmi codec is not supported */
+	for (i = HDMI_LINK_START; i <= HDMI_LINE_END; i++) {
+		link = &card->dai_link[i];
+		link->codecs[0].name = "snd-soc-dummy";
+		link->codecs[0].dai_name = "snd-soc-dummy-dai";
+	}
+}
+#endif
+
+static int snd_ehl_rt5660_probe(struct platform_device *pdev)
+{
+	struct snd_soc_acpi_mach *mach;
+	struct snd_soc_card *card = &snd_soc_card_ehl_rt5660;
+	struct sof_card_private *ctx;
+	int ret;
+
+	card->dev = &pdev->dev;
+
+	ctx = devm_kzalloc(&pdev->dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+	INIT_LIST_HEAD(&ctx->hdmi_pcm_list);
+#endif
+	snd_soc_card_set_drvdata(card, ctx);
+
+	mach = pdev->dev.platform_data;
+	ret = snd_soc_fixup_dai_links_platform_name(card,
+						    mach->mach_params.platform);
+	if (ret)
+		return ret;
+
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+	sof_refine_hdmi_links(card, mach);
+#endif
```

Comment:
@libinyang I was wondering what happens if CONFIG_SND_HDA_CODEC_HDMI is not set? The tplg file for this platform has the iDisp links, so the card probe will fail in this case, wouldn't it?

I think you could just put the ifdef check inside sof_redefine_hdmi_links() and if the hdmi codec driver is not built in, you replace the links with dummy codec and topology load will still work.

Reply:
@kv2019i If CONFIG_SND_HDA_CODEC_HDMI is not defined, there is no such dai links defined. When we define the hdmi dai links, we use:
```
#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
#define HDMI_LINK_START 3
#define HDMI_LINE_END 6
    {
        .name = "iDisp1",
        .id = 5,
 ...
#endif
```

Reply:
@libinyang hmm, but the topology still has the HDMI iDisp links, right? In that case topology load will fail if CONFIG_SND_HDA_CODEC_HDMI is not defined. We don't have logic to pick a topology that does not have HDMI.

Reply:
@kv2019i For EHL, we have different tplgs. If the kernel doesn't support hdmi, we have to pick the tplg without hdmi. This is the way our current driver does. Maybe we can add a feature that dynamically pick the tplg based on the kernel config. What do you think? It should be easy to implement and I would like do it later : - )

Reply:
@kv2019i @libinyang Wouldn't it be easier to select HDMI in Kconfig? Can there be any good reason not to want it? What are the disadvantages of forcing it on?

Reply:
@libinyang But this is problematic as tplg needs to be aligned to Kconfig manually. What I did in HDA generic driver is this: https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/skl_hda_dsp_generic.c#L153
... if SOF is built without HDMI, or codec driver is not built, we use the dummy codec but code probe still works and can still luse the same topology. I.e. you get what @lyakh you comment, if you disale HDMI in Kconfig (but keep same FW and topology), driver will continue to work, just without HDMI.

Reply:
> @kv2019i @libinyang Wouldn't it be easier to select HDMI in Kconfig? Can there be any good reason not to want it? What are the disadvantages of forcing it on?

@lyakh I think we should still let customer select to use intel's HDMI audio or not. Maybe customer won't use intel gfx.

Reply:
> @libinyang But this is problematic as tplg needs to be aligned to Kconfig manually. What I did in HDA generic driver is this: https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/skl_hda_dsp_generic.c#L153
> ... if SOF is built without HDMI, or codec driver is not built, we use the dummy codec but code probe still works and can still luse the same topology. I.e. you get what @lyakh you comment, if you disale HDMI in Kconfig (but keep same FW and topology), driver will continue to work, just without HDMI.

@kv2019i hmm.. I was referring the latest code of sof_pcm512x and sdw_rtxxx, which is sof_sdw.c now. They won't allocate hdmi dai link if HDMI is not enabled in Kconfig and I was thinking this is our recent recommended method. I will check skl_hda_dsp_generic.c.

Reply:
> > @kv2019i @libinyang Wouldn't it be easier to select HDMI in Kconfig? Can there be any good reason not to want it? What are the disadvantages of forcing it on?
> 
> @lyakh I think we should still let customer select to use intel's HDMI audio or not. Maybe customer won't use intel gfx.

@libinyang you mean they don't use the HDMI output from the i915? Sure, that's up to the user, but what negative effects does it have on the user? That they will have additional PCMs in their lists? That's what always happens - you usually have more entries there than you use, that shouldn't be a problem?

Reply:
Yes, I mean users may use Nvidia, for example, for output. In this case, intel HDMI audio initialization will fail. In legacy hda, we will not load Intel HDMI audio and not create the HDMI PCMs, but only load Nvidia HDMI audio. Now, we will keep the PCMs and make them as dummy codec, as Kai suggested. Maybe we can refine the HDMI audio support policy to dynamically switch tplg. But this should be in another topic.